### PR TITLE
test: add a parallel runner for the npm test chain

### PR DESCRIPTION
**Pull request by ulsklyc**
## Motivation

`npm test` is one `&&` chain of 327 steps (325 at the measured commit below). It runs serially, and the first red step hides every step after it: whoever breaks three suites finds out in three runs. This is the first building block of the test consolidation (R5): a runner that runs the same chain side by side, keeps going after a failure and says what was red.

`npm test` and CI are untouched. The chain stays the one registry of suites.

## What changes

- **`scripts/run-tests-parallel.mjs`**, started as `npm run test-parallel`. Every step of `scripts.test` runs as its own process (`/bin/sh -c <step>`, own process group), by default `os.availableParallelism() - 1` at a time, `--jobs N` to change it. One log per step, output through a file descriptor rather than a pipe, exit code straight from the child. It keeps going after a red step and ends with the red steps and their log paths, total wall time, the ten slowest steps, a `summary.json` in the log directory and exit 1. Further options: `--logs DIR`, `--timeout SECONDS` (default 900 per step, kills the process group), `--list`, `--package FILE`.
- **Location:** `scripts/` holds the Node scripts a developer starts through npm (`check-release-cadence.mjs`, `seed-demo.js`); `tools/` holds what ships to users (installer, quadlet).
- **Name:** a runner is not a suite. `test:suite-chain` demands every `test:*` script inside a chain and names the one entry that cannot sit inside itself (`test:document-guards`) exactly once, on purpose not as a growing list. The same rule that keeps `coverage` (which also only runs the chain) out of `test:` applies here, so no second name exception: `test-parallel`.
- **No second list:** the steps are parsed from `scripts.test`. The parser splits only at `&&` outside quotes and aborts on any other operator (`||`, `;`, `|`, `&`, parentheses) and on `$(`, `${` or backticks anywhere in a step, quoted or not, instead of letting it slide into a neighbouring step.
- **`test:parallel-runner`** (new suite, registered in the chain after `test:suite-chain`) holds parity against `sh` itself: the whole chain and every parsed step alone run against recording stubs of `node` and `npm` that log program, arguments and environment. Both records must match in count, order and content, so a lost `DB_PATH=:memory:` counts like a lost step. The real chain is splittable at ` && ` today, so the comparison alone would prove nothing; the suite also runs a chain with `&&` inside quotes, env prefixes and `2>&1`, where a naive split, a lost last step and two glued steps must turn red. It also runs the runner as a program: a red step does not stop the run, the exit code comes from the child, a hanging step falls to `--timeout`, a bad call exits 2.
- **Counterproof of the parity suite** (runner patched from a `cp` backup, restored and compared with `cmp`): parser drops the last step - 4 of 7 tests red; parser is `split(' && ')` - 2 of 7 red (the two tests built for it; the real-chain comparison stays green, as its comment says).
- Docs: `CONTRIBUTING.md` (Running tests) and `docs/test-suites.md`. No CHANGELOG entry: test tooling only, like #1223.

## After review

Review rounds (own review, Codex, claude-review) changed the runner after the measurement below:

- **Logs:** without `--logs` every run gets its own directory from `mkdtempSync` directly under the temp directory (owner-only, unpredictable name). Concurrent runs never share one; there is no fixed root and nothing is chmod-ed. Cleanup only removes the caller's own real directories (`lstat`, uid) older than 24 h.
- **Interrupt:** after SIGINT/SIGTERM no new step starts, every process group gets SIGTERM and after `--grace` (default 5 s) SIGKILL, then exit 130/143.
- **Timeout and interrupt checks** read the process state (`scripts/process-state.mjs`: `/proc/<pid>/stat` on Linux, `ps` elsewhere), so a zombie that a container init has not reaped counts as terminated.
- **Main guard** compares `realpathSync` of both paths; started through a symlink the runner no longer exits 0 silently.
- **`test:backup-scheduler`** uses its own temp directory per process instead of `./test-backups` (two copies at once were red 9 of 10 times before).
- Every change has a test that was red against the previous state. On `a7d4c953`: `npm run test-parallel` 327/327 green in 2m34s (7 jobs), with no other checkout's tests running.

## Measurement

Same commit (`8677dd8a`), MacBook with 8 cores / 8 GB, Node 26.8.2. Before every full run `pgrep -fl "test/test-"` checked for test processes of other checkouts, with a 30 s wait loop: no wait before the serial run, the first runner run and the 14-job run; about 2 min before run 2, 30 s each before run 3 and the shuffled run. A snapshot right after the last run showed no test process left behind by the runner, so those waits were most likely other checkouts' runs.

| Run | Jobs | Wall time | Result |
|---|---|---|---|
| `npm test` (serial) | 1 | 5m27s | exit 0 |
| `npm run test-parallel`, run 1 | 7 (default) | 1m54s | 325/325 green |
| `npm run test-parallel`, run 2 | 7 (default) | 1m50s | 325/325 green |
| `npm run test-parallel`, run 3 | 7 (default) | 2m09s | 325/325 green |
| shuffled order (fixed seed) | 7 | 2m00s | 325/325 green |
| `npm run test-parallel -- --jobs 14` (stress) | 14 | 1m55s | 325/325 green |

An exploratory run on the same code before the commit took 1m21s (7 jobs, 325/325 green); the later runs were slower, the machine was not otherwise idle. `test:parallel-runner` itself costs about 8 s alone (about 650 short `sh` calls), about 20 s under the runner.

The sum of the step durations under the runner is higher than the serial wall time (contention and one `npm` start per step), the wall time drops to roughly a third.

## Shared resources - inventory and fix

Inventoried by grep (fixed ports, fixed paths, temp dirs, DB paths, git calls, `chdir`), then checked with the full runs above.

| Resource | Where | Finding | Action |
|---|---|---|---|
| Fixed port 39871, probe port 39872 | `test/test-backup-webdav.js` | Real collision: two concurrent copies of the old suite, one aborted with exit 134. On Linux both ports lie in the ephemeral range (32768-60999). | **Fixed at the source:** one mock host on a kernel-assigned port for the whole suite (the service reads `WEBDAV_BACKUP_URL` once at import), each block swaps its handler; the unreachable probe takes a just-released port. Two concurrent copies of the new suite: both green; serially green (19/19). |
| PID-named temp DBs | `freshTestDbPath()` users | PIDs are unique among live processes; reuse over time is already handled by removing the file before opening. | none |
| `process.env.PORT = '3100'` | `test/test-mcp.js` | Only composes an expected URL, `fetch` is mocked, nothing listens. | none |
| Port literals 3000/8000/8080/8090/9925 | SSRF, push, notification, installer suites | URL strings for validators, never bound. | none |
| `/tmp/yuvomi-drive-override` | `test/test-document-storage.js` | Env value for a config check, never created. | none |
| `./test-backups` | `test/test-backup-scheduler.js` | Relative to the checkout, used by this one suite only. | none |
| Repo-root `yuvomi.db` | any suite without `DB_PATH` | Held by `test:db-isolation`; no stray file after the runs. | none |
| git | `test:migrations-append-only`, `test:changelog` | Read-only (`merge-base`, `show`, `tag --list`). | none |
| Order of the first steps (`test-db.js`, `schema-reconcile`, ...) | chain | Every suite opens its own in-memory or temp DB. A run in shuffled order (fixed seed, 7 jobs, `test:split-expenses-attachments` and `test:bind-address` first) stayed 325/325 green. | none |

## Serial allowlist

`SERIAL` in the runner is **empty**. Nothing had to run alone once the fixed port was gone. The list stays in place with its rule: an entry needs a reason, and an entry that no longer names a chain step aborts the run (and fails `test:parallel-runner`).

## Open questions

- **CI parallel?** A separate decision. `ubuntu-latest` has 4 cores, the matrix runs two Node versions; the per-step logs would need to become an artifact. This PR leaves `ci.yml` alone.
- Is 900 s per step the right default timeout? The serial chain has none; the slowest step measured here (`test:frontend-audit`) took at most 45 s under 7 jobs.
- Name: `test-parallel` next to `coverage`, or rather a `check:` name like `check:waste`?

### CODE REVIEW

File: test/test-parallel-runner.js [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+});
+
+/** Lebt `pid` noch? Wartet bis zu `ms`, weil ein getoeteter Waise erst vom Init-Prozess abgeraeumt wird. */
+function aliveAfter(pid, ms) {
+  const tick = new Int32Array(new SharedArrayBuffer(4));
+  for (const deadline = Date.now() + ms; ; Atomics.wait(tick, 0, 0, 25)) {
+    try { process.kill(pid, 0); } catch { return false; }
+    if (Date.now() > deadline) return true;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply test/test-parallel-runner.js]: Fixed in b63c46b5: scripts/process-state.mjs reads the process state (/proc/<pid>/stat on Linux, ps elsewhere), so a zombie counts as terminated; tested with a real zombie and with both readers stubbed, red again once kill(pid, 0) is used.

File: scripts/run-tests-parallel.mjs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+  let logDir = opts.logs;
+  if (!logDir) {
+    // Ein fester Ordner je Checkout, vor jedem Lauf geleert. Ein neuer Ordner
+    // je Lauf liess rund 326 Dateien pro Lauf im Temp-Ordner liegen; alte
+    // Ordner nach Praefix wegzuraeumen traefe den laufenden Runner eines
+    // anderen Arbeitsbaums.
+    const checkout = realpathSync(cwd);
+    const key = createHash('sha256').update(checkout).digest('hex').slice(0, 8);
+    logDir = path.join(os.tmpdir(), 'yuvomi-test-parallel', `${path.basename(checkout)}-${key}`);
+    rmSync(logDir, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed in b63c46b5 and a7d4c953: without --logs every run gets its own mkdtemp directory directly under tmpdir, so concurrent runs never share or delete each other's logs; a test starts two runs and checks both directories.

File: scripts/run-tests-parallel.mjs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+  }
+  mkdirSync(logDir, { recursive: true });
+
+  const stop = (signal) => {
+    for (const child of running) {
+      try { process.kill(-child.pid, 'SIGTERM'); } catch { /* schon weg */ }
+    }
+    console.error(`\nAbbruch (${signal}) - Logs in ${logDir}`);
+    process.exit(130);
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed in b63c46b5: after an interrupt no new step starts, every process group gets SIGTERM and after --grace (default 5 s) SIGKILL, then the runner exits 130/143; red against the old runner, where a SIGTERM-ignoring step and its grandchild survived.

File: scripts/run-tests-parallel.mjs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+  for (let i = 0; i < script.length; i += 1) {
+    const c = script[i];
+    if (quote === '\'') {
+      current += c;
+      if (c === '\'') quote = null;
+    } else if (quote === '"') {
+      current += c;
+      if (c === '\\' && i + 1 < script.length) current += script[++i];
+      else if (c === '"') quote = null;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs:133]: Fixed in b63c46b5: $(, backticks and ${ are rejected anywhere in a step, quoted or not, with exit 2 instead of splitting; your example is a test case and was split into three parts before.

File: scripts/run-tests-parallel.mjs [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    // Ein fester Ordner je Checkout, vor jedem Lauf geleert. Ein neuer Ordner
+    // je Lauf liess rund 326 Dateien pro Lauf im Temp-Ordner liegen; alte
+    // Ordner nach Praefix wegzuraeumen traefe den laufenden Runner eines
+    // anderen Arbeitsbaums.
+    const checkout = realpathSync(cwd);
+    const key = createHash('sha256').update(checkout).digest('hex').slice(0, 8);
+    logDir = path.join(os.tmpdir(), 'yuvomi-test-parallel', `${path.basename(checkout)}-${key}`);
+    rmSync(logDir, { recursive: true, force: true });
+  }
+  mkdirSync(logDir, { recursive: true });
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed in a7d4c953: the default log directory now comes from mkdtempSync (0700, unpredictable name) instead of a fixed path created with the umask.

File: scripts/run-tests-parallel.mjs [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+  privateDir(base);
+  const now = Date.now();
+  for (const entry of readdirSync(base, { withFileTypes: true })) {
+    if (!entry.isDirectory() || !entry.name.startsWith('run-')) continue;
+    const dir = path.join(base, entry.name);
+    try {
+      if (now - statSync(dir).mtimeMs > RUN_DIR_MAX_AGE_MS) rmSync(dir, { recursive: true, force: true });
+    } catch { /* raeumt gerade ein anderer Lauf weg */ }
+  }
+  return mkdtempSync(path.join(base, 'run-')); // mkdtemp legt mit 0700 an
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed in fb82609c: test-backup-scheduler uses its own temp directory per process instead of ./test-backups; two copies at once were red 9 of 10 times before and green 10 of 10 after. No other suite writes a checkout-relative directory.

File: scripts/run-tests-parallel.mjs [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+ * schrieben ihre summary.json an dieselbe Stelle (Review auf #1229).
+ * Aufgeraeumt werden nur Geschwister ueber 24 h - ein laufender Lauf ist
+ * frisch, denn jedes neue Log setzt die mtime seines Ordners.
+ */
+function defaultLogDir(cwd) {
+  const checkout = realpathSync(cwd);
+  const key = createHash('sha256').update(checkout).digest('hex').slice(0, 8);
+  const root = path.join(os.tmpdir(), 'yuvomi-test-parallel');
+  const base = path.join(root, `${path.basename(checkout)}-${key}`);
+  privateDir(root);
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Resolved differently in a7d4c953: there is no shared root any more. Each run gets a mkdtemp directory directly under tmpdir, owner-only, so another user's run can never block or read it.

File: scripts/run-tests-parallel.mjs [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+ * Die Logs tragen Testausgaben, und unter Linux liegt `/tmp` fuer alle Nutzer
+ * offen; mit der umask 022 entstanden Ordner 0755 und Dateien 0644.
+ */
+function privateDir(dir) {
+  mkdirSync(dir, { recursive: true, mode: 0o700 });
+  const stat = statSync(dir);
+  const own = typeof process.getuid !== 'function' || stat.uid === process.getuid();
+  if (own && (stat.mode & 0o777) !== 0o700) chmodSync(dir, 0o700);
+}
+
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed in a7d4c953: the fixed, predictable root is gone. Each run creates its directory atomically with mkdtempSync (0700, unpredictable name), nothing is chmod-ed, and cleanup only removes the caller's own real directories older than 24 h, checked with lstat and uid; a test plants a symlink in a substituted TMPDIR and was red against fb82609c.

File: scripts/run-tests-parallel.mjs [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+    } else if (c === '\\') {
+      current += c;
+      if (i + 1 < script.length) current += script[++i];
+    } else if (c === '\'' || c === '"') {
+      current += c;
+      quote = c;
+    } else if (c === '&') {
+      if (script[i + 1] === '&') {
+        push(i);
+        i += 1;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs:143]: Fixed up to 1d2ef8fc: an unquoted # that starts a word is rejected with exit 2 instead of being split around; a#b, a quoted # and \# stay allowed. The chain in package.json contains no #. The parser is not a shell and now says so in its header: anything beyond simple commands with env prefixes and quotes is rejected, not interpreted.

File: scripts/run-tests-parallel.mjs [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+
+function runStep({ step, index }, { cwd, logDir, timeoutMs }) {
+  const logPath = path.join(logDir, `${String(index + 1).padStart(3, '0')}-${slug(step)}.log`);
+  const fd = openSync(logPath, 'w');
+  const started = performance.now();
+  return new Promise((resolve) => {
+    let timedOut = false;
+    let timer = null;
+    // Eigene Prozessgruppe: ein Timeout oder Abbruch trifft auch npm -> sh -> node.
+    const child = spawn('/bin/sh', ['-c', step], { cwd, stdio: ['ignore', fd, fd], detached: true });
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs:375]: Fixed up to 1d2ef8fc: a step whose command is cd, export, unset, source, ., set, alias, ulimit, umask, pushd or popd, or that consists of assignments only, is rejected with exit 2, because each step runs in a fresh shell. Env prefixes in front of a command stay allowed; the real chain still lists 327 steps.

File: scripts/run-tests-parallel.mjs [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+      if (i + 1 >= argv.length) throw new Error(`${arg} braucht einen Wert`);
+      return argv[++i];
+    };
+    const count = (raw, min) => {
+      if (!/^\d+$/.test(raw) || Number(raw) < min) throw new Error(`${arg} erwartet eine ganze Zahl >= ${min}, nicht '${raw}'`);
+      return Number(raw);
+    };
+    if (arg === '--jobs' || arg === '-j') opts.jobs = count(value(), 1);
+    else if (arg === '--logs') opts.logs = path.resolve(value());
+    else if (arg === '--timeout') opts.timeout = count(value(), 0);
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs]: Fixed up to 1d2ef8fc: --timeout (and --grace) above 2147483 seconds is rejected with exit 2 instead of reaching an overflowing timer; --timeout 0 still means no limit.

File: scripts/run-tests-parallel.mjs [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+  try { names = readdirSync(tmpdir); } catch { return removed; }
+  for (const name of names.sort()) {
+    if (!name.startsWith(RUN_PREFIX)) continue;
+    const dir = path.join(tmpdir, name);
+    try {
+      const stat = lstatSync(dir);
+      if (stat.isSymbolicLink() || !stat.isDirectory()) continue;
+      if (typeof uid === 'number' && stat.uid !== uid) continue;
+      if (now - stat.mtimeMs <= maxAgeMs) continue;
+      rmSync(dir, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs:355]: Fixed in da5563eb: the runner refreshes its run directory's mtime at start, after every step and every 10 minutes, so a long quiet run is never older than the 24 h cleanup threshold; a test with a short heartbeat was red before (the cleanup removed the active run).

File: test/test-backup-webdav.js [hunk 13]
```diff
@@ -17,7 +17,59 @@ const xmlEsc = (s) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/
...
+}
+
+/** Ein Port, auf dem gerade niemand lauscht: vom Kernel geholt und wieder freigegeben. */
+function closedPort() {
+  const probe = http.createServer();
+  return new Promise((resolve, reject) => {
+    probe.once('error', reject);
+    probe.listen(0, '127.0.0.1', () => {
+      const { port } = probe.address();
+      probe.close(() => resolve(port));
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply test/test-backup-webdav.js]: Fixed in aaaee381: the transport failure now comes from 127.0.0.1:2 (port 1 is refused by fetch as a bad port before connecting), which gives ECONNREFUSED without reserving and releasing a port. Red before with a neighbour binding the released port; 20 copies at once are green now.

File: scripts/run-tests-parallel.mjs [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+    } else {
+      word += c;
+    }
+  }
+  if (word) words.push(word);
+  return words;
+}
+
+const ASSIGNMENT = /^[A-Za-z_][A-Za-z0-9_]*=/;
+const STATE_CHANGING = new Set(['cd', 'export', 'unset', 'source', '.', 'set', 'alias', 'ulimit', 'umask', 'pushd', 'popd']);
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply scripts/run-tests-parallel.mjs:194]: Not changed: the runner's parser is deliberately not a shell (see GRENZEN in its header), and further parser edge cases are answered by measurement rather than new rules. The test chain in package.json contains no exit or exec step (0 of each), and test:parallel-runner compares every parsed step against sh itself, so a chain that gained one would be looked at there.

File: test/test-backup-webdav.js [hunk 15]
```diff
@@ -17,7 +17,65 @@ const xmlEsc = (s) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/
...
+ * nachgestellt: "Missing expected rejection").
+ *
+ * Port 2 bekommt keine Suite: `listen(0)` vergibt nur ephemere Ports (unter
+ * Linux ab 32768, unter macOS ab 49152), und unter Linux darf ein Prozess ohne
+ * Root-Rechte ihn gar nicht binden. Port 1 geht NICHT: fetch sperrt ihn als
+ * "bad port" der Fetch-Spezifikation, bevor es eine Verbindung versucht - der
+ * Fehler hat dort keinen `cause.code` (gemessen 2026-09-15, Port 2 bis 5:
+ * ECONNREFUSED in wenigen Millisekunden).
+ */
+const REFUSED_PORT = 2;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply test/test-backup-webdav.js:67]: Not changed: binding port 2 needs root or CAP_NET_BIND_SERVICE, and no suite or tool in this repo listens there; listen(0) only hands out ephemeral ports. The earlier release-and-reconnect approach raced under the parallel runner itself, which this removes. A machine that deliberately serves port 2 will see this one test fail loudly, not pass silently.
